kythe/docs/rfc/2967.md

```diff
@@ -0,0 +1,64 @@
+1.  The relevant information should be in a form which is readily available for filtering and display.
+
+The end result is that build configuration is most tightly bound to anchors both semantically and practically.  This design also allows for including references to entities regardless of build configuration as well as filtering only references for a particular set.  Indexers for which build configuration is irrelevant may omit the fact.
+
+
+## Detailed Design {#detailed-design}
+
+
+### Graph {#graph}
+
+The Kythe graph will be augmented such that:
+
+
+1.  File nodes with a non-empty root VName field are explicitly considered to have been generated by the build.  If the contents of generated files differ with the build configuration, the root must include a unique string per build configuration.  This follows from the general requirements that unique graph entities be given unique VNames.  This allows clients to trivially know if they should request the source for a particular file from the underlying repository or from Kythe.
+1.  Corpus represents a stable name to a particular logical body of code.  Different corpus names may map to different parts of the same underlying repository.  That mapping resides outside the Kythe graph[^1].  In particular, this can be used to accommodate source code which resides in a particularly repository, but is logically distinct from the primary corpus (perhaps due to using a different build system).  For example, within Piper the primary corpus is "google3" but the source code for several languages' standard library resides in third_party.  Presently, Java and C++ use distinct corpora for these references, but Go uses a separate root.  Go will be changed to use the "gostdlib" corpus for this.
+
+
+1.  A new fact "build/config" will be allowed on anchor nodes.  This should be a user-meaningful string representation of all relevant build characteristics (think "android" not "-DHAVE_GLIB").  
+1.  As we expect distinct anchors at the same position, this fact must be incorporated into the anchor signature by the indexer when present.
+
+
+### Extraction and Indexing {#extraction-and-indexing}
+
+As build configuration is a high-level description of a build, but must be incorporated into anchors by the indexer, it must be provided or derived at build time and incorporated into each resulting CompilationUnit.  This will be done by adding a string-value "build_config" field to the[ kythe.proto.BuildDetails](https://cs.corp.google.com/piper///depot/google3/third_party/kythe/kythe/proto/buildinfo.proto) message.  Indexers must incorporate this field into emitted anchors if the language in question makes a meaningful distinction between build configurations.  This is primarily relevant to C/C++ family languages, but is up to the individual language indexers to make this determination.  When in doubt, it is recommended that these behavior be flag controlled.
```

Comment:
It's probably worth mentioning somewhere that build configurations—like corpus labels—are assumed to have some kind of stability to them over time. This is kind of implied by the examples of platform, but it would bear saying it explicitly.

Reply:
Done.